File: windows/src/core/transcription/keytermBudget.ts
/**
 * Deepgram's `keyterm` prompting has a hard per-request token limit.
 *
 * Exceeding it is not a soft failure. The API rejects the entire request with
 * `Keyterm limit exceeded. The maximum number of tokens across all keyterms is
 * 500.` — so an oversized personal dictionary breaks *every* dictation, not just
 * the one with the long term list.
 *
 * The token count is produced by Deepgram's own tokenizer, which cannot be run
 * locally, so this estimates and stays well clear of the ceiling.
 * https://developers.deepgram.com/docs/keyterm#key-term-limits
 */
export const HARD_TOKEN_LIMIT = 500;

/**
 * Headroom. Real tokenization differs from any local estimate (punctuation,
 * unusual casing, digits, non-ASCII scripts), and Deepgram's own guidance is to
 * stay "well under" the limit and focus on the most important 20-50 terms.
 */
export const SAFETY_MARGIN_TOKENS = 100;

/** The token budget this app is willing to spend on keyterms in one request. */
export const TOKEN_BUDGET = HARD_TOKEN_LIMIT - SAFETY_MARGIN_TOKENS;

/**
 * Upper bound on how many keyterm strings are sent, regardless of token cost.
 *
 * Every keyterm is URL-encoded onto the request line, so an unbounded list also
 * risks a URL that an intermediary proxy rejects.
 */
export const MAX_TERMS = 100;

/** A keyterm so long it can only be a pasted paragraph, not a dictionary term. */
export const MAX_TERM_LENGTH = 64;

/**
 * Conservative per-term token estimate.
 *
 * Deliberately over-estimates. Deepgram tokenizes into subword units, and
 * technical terms, CamelCase identifiers, digits and non-Latin scripts all split
 * much more aggressively than plain lowercase English. Under-counting is the
 * dangerous direction: it would let the list pass this check and then be rejected
 * by the API.
 */
export function estimateTokens(term: string): number {
  const trimmed = term.trim();
  if (trimmed.length === 0) return 0;

  const words = trimmed.split(/\s+/).filter((word) => word.length > 0).length;
  const separators = (trimmed.match(/[^\p{L}\p{N}]/gu) ?? []).length;

  // Per-script cost. The previous version assumed ~5 characters per token for
  // every script, which undercounts CJK by roughly 5x: a 5-character Japanese
  // term was estimated at 1 token when its real cost is closer to 5. Japanese and
  // Chinese are reachable through the `auto` language pin, so a CJK dictionary
  // could blow the ceiling while the estimator reported a comfortable margin.
  let lengthTokens = 0;
  for (const char of trimmed) {
    lengthTokens += tokenWeight(char);
  }
  const byLength = Math.ceil(lengthTokens / 5);

  return Math.max(1, Math.max(words, byLength) + separators);
}

/**
 * How many BPE tokens one character plausibly costs, in units of "average Latin
 * characters per token" (~5 for the conservative end of common vocabularies).
 * Dividing by 5 afterwards keeps one consistent currency.
 */
function tokenWeight(char: string): number {
  const code = char.codePointAt(0) ?? 0;
  // Han, Hiragana, Katakana, Hangul: roughly one token per character.
  if (
    (code >= 0x3040 && code <= 0x30ff) || // kana
    (code >= 0x3400 && code <= 0x4dbf) || // CJK ext A
    (code >= 0x4e00 && code <= 0x9fff) || // CJK unified
    (code >= 0xf900 && code <= 0xfaff) || // CJK compat
    (code >= 0xac00 && code <= 0xd7af) || // Hangul syllables
    (code >= 0x1100 && code <= 0x11ff)    // Hangul jamo
  ) {
    return 5;
  }
  // Cyrillic, Greek, Arabic, Hebrew, Thai, Devanagari: split more than Latin.
  if (code >= 0x0370 && code <= 0x0fff) return 3;
  if (code >= 0x0e00 && code <= 0x109f) return 3;
  return 1;
}

/** Total estimated cost of a keyterm list. */
export function estimateListTokens(terms: readonly string[]): number {
  return terms.reduce((total, term) => total + estimateTokens(term), 0);
}

/** True when a list would exceed the budget this app is willing to spend. */
export function exceedsBudget(terms: readonly string[]): boolean {
  return terms.length > MAX_TERMS || estimateListTokens(terms) > TOKEN_BUDGET;
}

/**
 * Reject entries the provider cannot use, before they reach a request: empty
 * strings, punctuation-only strings, control characters (which would corrupt the
 * query string), and absurdly long values.
 */
export function isSendableKeyterm(term: string): boolean {
  const trimmed = term.trim();
  if (trimmed.length === 0 || trimmed.length > MAX_TERM_LENGTH) return false;
  if (!/[\p{L}\p{N}]/u.test(trimmed)) return false;
  // eslint-disable-next-line no-control-regex
  return !/[\u0000-\u001f\u007f]/u.test(trimmed);
}

File: Sources/UsefulVoiceCore/Transcription/KeytermBudget.swift
import Foundation

/// Deepgram's `keyterm` prompting has a hard per-request token limit. Exceeding
/// it is not a soft failure: the API rejects the whole request with
/// `Keyterm limit exceeded. The maximum number of tokens across all keyterms is 500.`
/// so a large personal dictionary would break *every* dictation.
/// https://developers.deepgram.com/docs/keyterm#key-term-limits
///
/// The token count is measured by Deepgram's own tokenizer, which we cannot run
/// locally, so we estimate and stay well under the ceiling.
public enum KeytermBudget {
    /// Deepgram's documented hard ceiling: 500 tokens across all keyterms.
    public static let hardTokenLimit = 500

    /// Reserve headroom. Real tokenization differs from our estimate (punctuation,
    /// unusual casing, digits, non-ASCII), and Deepgram's guidance is to stay
    /// "well under" the limit and focus on the top 20-50 terms.
    public static let safetyMarginTokens = 100

    /// The token budget we are willing to spend on keyterms in one request.
    public static var tokenBudget: Int { hardTokenLimit - safetyMarginTokens }

    /// Upper bound on how many keyterm strings we send regardless of token cost.
    /// Keeps the query string (and therefore the request URL) short: every term
    /// is URL-encoded into the request line, and servers commonly reject request
    /// URIs beyond a few kilobytes.
    public static let maxTerms = 100

    /// Conservative per-term token estimate.
    ///
    /// Deepgram tokenizes words into subword units, so the estimate must never
    /// undercount. We take the larger of "word count" and a character-length
    /// ratio: technical terms (`Kubernetes`, `TypeScript`, `fine-tune`) and
    /// hyphenated or CamelCase strings frequently split into several subword
    /// tokens. Digits also split aggressively (`GPT-4` -> `GPT`, `-`, `4`).
    public static func estimatedTokens(for term: String) -> Int {
        let trimmed = term.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return 0 }

        let words = trimmed.split(whereSeparator: { $0.isWhitespace }).count
        // ~5 characters per subword token is the conservative end of common BPE
        // vocabularies for Latin script; round up so we never undercount.
        // Non-Latin scripts split far more aggressively per character — kana,
        // Han and Hangul cost roughly one token each — so the previous flat
        // ~5-chars-per-token assumption undercounted them by roughly 5x. CJK is
        // reachable via the `auto` language pin, so a CJK dictionary could
        // exceed the ceiling while the estimate still looked safe. Weigh each
        // Unicode scalar (a code point, not a grapheme cluster) by script.
        var lengthTokens = 0
        for scalar in trimmed.unicodeScalars {
            lengthTokens += Self.tokenWeight(scalar)
        }
        let byLength = Int((Double(lengthTokens) / 5.0).rounded(.up))

        // Punctuation and internal separators are usually their own tokens.
        let separators = trimmed.filter { !$0.isLetter && !$0.isNumber }.count
        let base = max(words, byLength) + separators

        return max(1, base)
    }

    /// How many BPE tokens one Unicode scalar plausibly costs, in units of
    /// "average Latin characters per token" (~5 for the conservative end of
    /// common vocabularies). Dividing by 5 afterwards keeps one consistent
    /// currency.
    private static func tokenWeight(_ scalar: Unicode.Scalar) -> Int {
        let code = scalar.value
        // Han, Hiragana, Katakana, Hangul: roughly one token per character.
        if (code >= 0x3040 && code <= 0x30FF)      // kana
            || (code >= 0x3400 && code <= 0x4DBF)  // CJK ext A
            || (code >= 0x4E00 && code <= 0x9FFF)  // CJK unified
            || (code >= 0xF900 && code <= 0xFAFF)  // CJK compat
            || (code >= 0xAC00 && code <= 0xD7AF)  // Hangul syllables
            || (code >= 0x1100 && code <= 0x11FF) { // Hangul jamo
            return 5
        }
        // Cyrillic, Greek, Arabic, Hebrew, Thai, Devanagari, Myanmar etc.:
        // split more than Latin.
        if code >= 0x0370 && code <= 0x0FFF { return 3 }
        if code >= 0x0E00 && code <= 0x109F { return 3 }
        return 1
    }

    /// Total estimated cost of a keyterm list.
    public static func estimatedTokens(for terms: [String]) -> Int {
        terms.reduce(0) { $0 + estimatedTokens(for: $1) }
    }

    /// True when a list would exceed the budget we are willing to spend. Used by
    /// tests and by the settings diagnostics to catch a regression early.
    public static func exceedsBudget(_ terms: [String]) -> Bool {
        terms.count > maxTerms || estimatedTokens(for: terms) > tokenBudget
    }

    /// A single keyterm string so long that it can never be worth sending: it
    /// would consume the entire budget on its own and is almost certainly a
    /// pasted paragraph rather than a dictionary term.
    public static let maxTermLength = 64

    /// Rejects pathological entries before they reach a request: empty strings,
    /// strings that are only punctuation, control characters (which would break
    /// the query string or the API's parsing), and absurdly long values.
    public static func isSendableKeyterm(_ term: String) -> Bool {
        let trimmed = term.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, trimmed.count <= maxTermLength else { return false }
        guard trimmed.contains(where: { $0.isLetter || $0.isNumber }) else { return false }
        return !trimmed.unicodeScalars.contains { scalar in
            scalar.value < 0x20 || scalar.value == 0x7F
        }
    }
}
